nifi-nar-bundles/nifi-oauth-bundle/nifi-oauth/src/main/java/org/apache/nifi/oauth/OAuth2ClientCredentialsGrantControllerService.java

```diff
@@ -0,0 +1,167 @@
+import org.apache.oltu.oauth2.common.exception.OAuthProblemException;
+import org.apache.oltu.oauth2.common.exception.OAuthSystemException;
+import org.apache.oltu.oauth2.common.message.types.GrantType;
+
+@Tags({ "oauth2", "client", "secret", "post"})
+@CapabilityDescription("POSTs the ClientId and ClientSecret to the OAuth2 authentication server to retrieve the" +
+        " access token. The access token is stored locally in the controller service and used by processors " +
+        "referencing this controller service.")
+public class OAuth2ClientCredentialsGrantControllerService
+        extends AbstractOAuthControllerService
+        implements OAuth2ClientService {
+
+    public static final PropertyDescriptor CLIENT_ID = new PropertyDescriptor
+            .Builder().name("OAuth2 Client ID")
+            .displayName("OAuth2 Client ID")
+            .description("OAuth2 Client ID passed to the authorization server")
+            .required(true)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    public static final PropertyDescriptor CLIENT_SECRET = new PropertyDescriptor
+            .Builder().name("OAuth2 Client Secret")
+            .displayName("OAuth2 Client Secret")
+            .description("OAuth2 Client Secret that will be passed to the authorization server in exchange for an access token")
+            .sensitive(true)
+            .required(true)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    private static final List<PropertyDescriptor> properties;
+
+    static {
+        final List<PropertyDescriptor> props = new ArrayList<>();
+        props.add(AUTH_SERVER_URL);
+        props.add(CLIENT_ID);
+        props.add(CLIENT_SECRET);
+        props.add(RESPONSE_ACCESS_TOKEN_FIELD_NAME);
+        props.add(RESPONSE_EXPIRE_TIME_FIELD_NAME);
+        props.add(RESPONSE_EXPIRE_IN_FIELD_NAME);
+        props.add(RESPONSE_SCOPE_FIELD_NAME);
+        props.add(RESPONSE_TOKEN_TYPE_FIELD_NAME);
+        properties = Collections.unmodifiableList(props);
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return properties;
+    }
+
+    private String clientId = null;
+    private String clientSecret = null;
+    private Base64.Encoder enc = Base64.getEncoder();
+
+    /**
+     * @param context the configuration context
+     * @throws InitializationException if unable to create a database connection
+     */
+    @OnEnabled
+    public void onEnabled(final ConfigurationContext context) throws InitializationException {
+
+        super.onEnabled(context);
+
+        clientId = context.getProperty(CLIENT_ID).getValue();
+        clientSecret = context.getProperty(CLIENT_SECRET).getValue();
+    }
+
+    public boolean authenticate() {
+        HttpClient con = null;
+
+        try {
+
+            String base64String = enc.encodeToString((clientId + ":" + clientSecret).getBytes());
```

Comment:
Why create the Base64 encoded header as well as put the client ID and client secret in the request querystring? According to [OAuth.com :: Client Credentials](https://www.oauth.com/oauth2-servers/access-tokens/client-credentials/), the OAuth server will accept one or the other. Is there a way to know in advance which method is supported, and if both are supported, only one is necessary, correct?

Reply:
If the authorization server is requested over HTTP (as in the case of the example service http://term.ie), both the request parameters and the headers are transmitted in plaintext. We should either a) prevent OAuth from being used with HTTP servers or at least b) warn visibly when the secret will be transmitted in plaintext.

Reply:
From [Understanding OAuth2](http://www.bubblecode.net/en/2016/01/22/understanding-oauth2/):

> HTTPS
> OAuth2 requires the use of HTTPS for communication between the client and the authorization server because of sensitive data passing between the two (tokens and possibly resource owner credentials). In fact you are not forced to do so if you implement your own authorization server but you must know that you are opening a big security hole by doing this.

Reply:
At first I thought we would just force HTTPS in the connection string. After reading that paragraph from the OAuth2 documentation it almost seems to me like just a warning is what we should do? What do you think?

Reply:
I say we enforce HTTPS, and if someone wants to connect to an HTTP auth server, we do not let them.